Basic unit testing

kba opened the pull request:
This PR introduces basic unit testing with shell scripts. Every unit test shall be a shell script in a dedicated directory under `./test`. The tests are written using a tiny DSL in `./test/testlib.sh`. The output should be [TAP-compatible](https://testanything.org/).

zuphilip commented:
Can you elaborate how the tests here relate to the other tests `./tests/run.sh`? Should both test(s) directories stay in the future?

kba commented:
`./tests/run.sh` is just three lines at the moment:

```js
../hocr-check sample.html
../hocr-extract-images -p "words-from-test-%03d.png" -e "ocrx_word" tess.hocr
rm words-from-test*
```

It serves more as [smoke-testing](https://en.wikipedia.org/wiki/Smoke_testing_(software)) because it only executes the scripts.

On the other hand, if you run `./runner.sh` in `./test`, you get structured output like this:

```tap
#
# Testing ./hocr-pdf/test-hocr-pdf.sh
#
ok 1 - Executed: wget --quiet http://digi.bib.uni-mannheim.de/fileadmin/digi/445442158/hocr/445442158_0126.hocr
ok 2 - Executed: wget --quiet http://digi.bib.uni-mannheim.de/fileadmin/digi/445442158/max/445442158_0126.jpg
ok 3 - Not empty file: 445442158_0126.pdf
ok 4 - Executed: pdfgrep tribunali 445442158_0126.pdf
```

We should convert the files in `./tests` to proper unit tests.

zuphilip commented on README.md:
````diff
@@ -162,7 +163,11 @@ Merges the Dublin Core metadata into the hOCR file by encoding the data in its h
+Usage: hocr-pdf <imgdir>
+```
+
+Create a searchable PDF from a pile of hOCR and JPEG
````
We already have this from the last commit: https://github.com/tmbdev/hocr-tools/commit/cbac14b4a3a5f5d37c73a19f4e6a4dad875ca3f5

zuphilip commented on README.md:
````diff
@@ -172,3 +177,41 @@ hocr-split file.html pattern
+
+```sh
+#!../runner.sh
+
````
I guess this needs some update for the new `tsht`

kba replied:
Indeed, I'll rebase (see above) and update the README.

zuphilip replied:
Okay, let me know when you are through rebasing. Is the PR ready for review?

zuphilip commented on README:
```diff
@@ -0,0 +1,73 @@
+Tools for manipulating and evaluating the hOCR microformat for
+representing multi-lingual OCR results.
+
+hOCR is a format for representing OCR output, including layout
```
@kba This file was already replaced with the markdown file. Therefore this should be deleted from the PR. (Maybe check the commit https://github.com/tmbdev/hocr-tools/pull/10/commits/91daa590d9661d85da61d6242752fb1bc27f19f1 if there are any other unwanted side-effects.)

kba replied:
rebased the wrong head. I've squashed and rebased on master the whole thing now, let's see what CI says.

kba commented:
I rebased on master and squashed most commits. Tests [fail](https://travis-ci.org/UB-Mannheim/hocr-tools/builds/131052166) as expected now and we can start fixing them :smile:

zuphilip commented on README.md:
````diff
@@ -162,11 +169,7 @@ Merges the Dublin Core metadata into the hOCR file by encoding the data in its h
-```
-
-Create a searchable PDF from a pile of hOCR and JPEG
+TODO
````
Please undo the deletion of these lines.

kba replied:
Wow, those four lines are getting on my nerves :anger:  1490e8a

zuphilip commented on README.md:
```diff
@@ -176,3 +179,47 @@ hocr-split file.html pattern
+take a look at the existing [unit tests](./test/).
+
+1) Create a new directory under `./test`
+2) Copy any test assets (images, hocr files...) to this directory
```
I guess we can also share the same images, hocr among multiple tests. Or is there any reason to restrict such that they have to be in the same directory?

kba replied:
No, they can be shared, it's just easier to keep tests self-contained when code and assets are in the same place. Otherwise you run the risk of having some content that is not actually used for tests.

zuphilip commented on test/legacy/sample.html:
```diff
@@ -0,0 +1,56 @@
+    <meta name='DC.creator' value='James T. Kirk'>
+    <meta name='DC.title' value='Nomad'>
+    <meta name='DC.publisher' value='NASA'>
+</head>
```
Why do we duplicate this file here again? Can we not use the same copy of the hocr file as in the test `test/hocr-check`?

kba replied:
Since I don't know the original intention of the `tests` folder, I kept the `legacy` folder as it was, only changed the shell script to test something. We also don't use the `dcsample.xml` `dcsample2.xml` files.

zuphilip commented on test/smoke.tsht:
```diff
@@ -0,0 +1,6 @@
+for f in check combine eval eval-geom eval-lines extract-g1000 extract-images lines merge-dc pdf split;do
+    exec_ok "hocr-$f" "--help" "hocr-$f --help"
+    exec_ok "hocr-$f" "-h" "hocr-$f -h"
+done
```
Also add the command without any argument here?

kba replied:
That won't work for commands that listen on STDIN, like hocr-check

zuphilip replied:
Agreed

zuphilip commented on test/tsht:
```diff
@@ -0,0 +1,22 @@
+
+# Export TSHTLIB and execute the test-runner
+export TSHTLIB
+exec "$TSHTLIB/tsht-runner.sh" "$@"
```
Can you check here that there is a newline at the end?

kba replied:
There is now (081c42e)

zuphilip commented on README.md:
```diff
@@ -176,3 +179,47 @@ hocr-split file.html pattern
+#!/usr/bin/env tsht
+
+# adjust to the number of your tests
+plan 1
```
This is not always given in the tests below. I would replace the comment with `# number of test statements in the following` or something like that. Is it important to add this line?

kba replied:
No, it is not required to plan ahead but it makes it easier to spot when a test just aborts.

zuphilip commented on test/hocr-pdf/test-hocr-pdf.tsht:
```diff
@@ -0,0 +1,11 @@
+exec_ok wget --quiet "http://digi.bib.uni-mannheim.de/fileadmin/digi/445442158/max/${work}.jpg"
+hocr-pdf . > "${work}.pdf"
+file_not_empty "${work}.pdf"
+exec_ok pdfgrep 'tribunali' "${work}.pdf"
```
This fails in travis which you linked:
```
not ok 1 - Test requires 'pdfgrep' and 'wget' installed
...
not ok 5 - Failed: pdfgrep tribunali 445442158_0126.pdf
```

kba replied:
Yes, the test could be cleverer. But to install pdfgrep in travis, we'd need sudo and that would mean no container-based infrastructure for Travis CI which slows down the testing.

kba replied:
The test now stops when it notices this.

kba replied:
The test now stops when it notices this.

zuphilip commented on .travis.yml:
```diff
@@ -0,0 +1,7 @@
+install:
+  - pip install -r requirements.txt
+  - python setup.py install
+  - npm install -g tap-spec
```
Where do we need `tap-spec` now?

kba replied:
I temporarily disabled it because it made tests that failed appear like they passed (tap-spec expects failed tests to exit 1 and only 1, not, as tsht does by default, with a non-zero value equal to the number of failed tests). This was fixed by https://github.com/kba/tsht/commit/352dcd7e5a3831996fe212c91d38d4d6478d1aa5, I'll reenable it.

zuphilip commented on .travis.yml:
```diff
@@ -0,0 +1,8 @@
+  - pip install pep8
+script: 
+  - ./test/tsht
+  - pep8 hocr-*
```
`setup.py` should also be added to check. But I really would suggest to look into this afterwards. Testing currently for this coding standard will just produce a lot of errors and warnings...

kba replied:
We can comment out the line. How to proceed with this PR?

zuphilip replied:
I will try to look and test it further in the next days, also your other PRs. I guess the order in which these PR are merged will not depend that much, right?

zuphilip commented on README.md:
````diff
@@ -176,3 +183,47 @@ hocr-split file.html pattern
+e.g.
+
+```sh
+./test/tsht hocr-pdf/test-hocr-pdf.tsht
````
We need the `test` directory also in the argument, i.e.
```sh
./test/tsht test/hocr-pdf/test-hocr-pdf.tsht
```

zuphilip commented:
Okay, I will merge that PR now. Everything looks fine and I tested the test framework locally. @kba Thank you very much for the work and your patience!

kba commented:
:tada: yay, great. I've just made another test for dc-merge, I'll open a new PR.
